### PR TITLE
using nearText or nearVector to request weaviate

### PR DESCRIPTION
**Pull request by ZouhairElhadi**
I updated the class Weavaite in the module weaviate, by:
- Adding an attribute by_text, takes the value True by default, if this attribute is true, we can use the nearText to request data from weaviate, else we use nearVector.
- I changed the name of similarity_search to similarity_search_by_text, which using nearText
- I added a new method similarity_search, which checks the attribute by_text and based on its value we can using nearText or nearVector

### CODE REVIEW

**Comment by dev2049** [comment]
hey @ZouhairElhadi, thanks for contribution! what's the use case you're imagining? why have an attribute on the VectorStore that determines which similarity search to use versus just letting your application call one or the other method?

**Comment by ZouhairElhadi** [comment]
> hey @ZouhairElhadi, thanks for contribution! what's the use case you're imagining? why have an attribute on the VectorStore that determines which similarity search to use versus just letting your application call one or the other method?

While attempting to utilize data stored in Weaviate using Haystack with a schema that includes the attribute "vectorizer":"none", an error occurred. The error message stated: "ValueError: Error during query: [{'locations': [{'column': 24, 'line': 1}], 'message': 'Unknown argument "nearText" on field "Document" of type "GetObjectsObj". Did you mean "nearVector" or "nearObject"?', 'path': None}]."
To determine which method to use for similarity calculations, we require the "by_text" attribute. By default, this attribute is set to "true," meaning that previous work will not be affected.

**Comment by hwchase17** [comment]
> > hey @ZouhairElhadi, thanks for contribution! what's the use case you're imagining? why have an attribute on the VectorStore that determines which similarity search to use versus just letting your application call one or the other method?
> 
> While attempting to utilize data stored in Weaviate using Haystack with a schema that includes the attribute "vectorizer":"none", an error occurred. The error message stated: "ValueError: Error during query: [{'locations': [{'column': 24, 'line': 1}], 'message': 'Unknown argument "nearText" on field "Document" of type "GetObjectsObj". Did you mean "nearVector" or "nearObject"?', 'path': None}]." To determine which method to use for similarity calculations, we require the "by_text" attribute. By default, this attribute is set to "true," meaning that previous work will not be affected.

can you clarify more how you are using? this doesn't sound like a super typical use case and we dont want to bloat the interface

**Review by khu834 — COMMENTED** [review]
would be great to also implement the same for `similarity_search_by_text_with_score`

**Comment by khu834** [comment]
> 

I have a similar use case too.
In order to use the Weaviate internal embedding model, you must define one in the schema like in the example.

```
schema = {
    "classes": [
        {
            "class": "Paragraph",
            "description": "A written paragraph",
            "vectorizer": "text2vec-openai",
              "moduleConfig": {
                "text2vec-openai": {
                  "model": "ada",
                  "modelVersion": "002",
                  "type": "text"
                }
              },
            "properties": [
                {
                    "dataType": ["text"],
                    "description": "The content of the paragraph",
                    "moduleConfig": {
                        "text2vec-openai": {
                          "skip": False,
                          "vectorizePropertyName": False
                        }
                      },
                    "name": "content",
                },
            ],
        },
    ]
}

client.schema.create(schema)
```
If you do not specify `vectorizer`, it will not allow you to use `nearText`.
In my case, I want to "Bring my own vectorizer" instead of going through weaviate and specifying it in the weaviate schema, so I would love to be able to use my own embedding function in the `similarity_search`

I think a good behavior could be that the one in the schema can be the default. However, if one isn't set, you can use a provided embedding function to do the job too, all within the same `similarity_search` function, without needing an additional variable.

File: langchain/vectorstores/base.py [hunk 1]
```diff
@@ -107,6 +107,12 @@ def similarity_search(
     ) -> List[Document]:
         """Return docs most similar to query."""
 
+    @abstractmethod
+    def similarity_search_by_text(
```
Comment (dev2049) [inline langchain/vectorstores/base.py:111]: could we remove this off base class? don't think this generalizes to all vector stores

**Review by dev2049 — COMMENTED** [review]
one comment, otherwise think it looks good. @hsm207 @hwchase17 do we think this resolves #4742

**Comment by hsm207** [comment]
> one comment, otherwise think it looks good. @hsm207 @hwchase17 do we think this resolves #4742

@dev2049 yes, this resolves #4742

**Comment by dev2049** [comment]
refactored slightly and landed in #4824. thanks @ZouhairElhadi!

**Pull request closed without merging**